src/native/minipal/cpufeatures.h

```diff
@@ -26,20 +26,12 @@ enum XArchIntrinsicConstants
     XArchIntrinsicConstants_Lzcnt = 0x1000,
     XArchIntrinsicConstants_AvxVnni = 0x2000,
     XArchIntrinsicConstants_Movbe = 0x4000,
-    XArchIntrinsicConstants_Avx512f = 0x8000,
-    XArchIntrinsicConstants_Avx512f_vl = 0x10000,
-    XArchIntrinsicConstants_Avx512bw = 0x20000,
-    XArchIntrinsicConstants_Avx512bw_vl = 0x40000,
-    XArchIntrinsicConstants_Avx512cd = 0x80000,
-    XArchIntrinsicConstants_Avx512cd_vl = 0x100000,
-    XArchIntrinsicConstants_Avx512dq = 0x200000,
-    XArchIntrinsicConstants_Avx512dq_vl = 0x400000,
-    XArchIntrinsicConstants_Avx512Vbmi = 0x800000,
-    XArchIntrinsicConstants_Avx512Vbmi_vl = 0x1000000,
-    XArchIntrinsicConstants_Serialize = 0x2000000,
-    XArchIntrinsicConstants_Avx10v1 = 0x4000000,
-    XArchIntrinsicConstants_Avx10v1_V512 = 0x8000000,
-    XArchIntrinsicConstants_Evex = 0x10000000,
+    XArchIntrinsicConstants_Avx512 = 0x8000,
+    XArchIntrinsicConstants_Avx512Vbmi = 0x10000,
+    XArchIntrinsicConstants_Serialize = 0x20000,
+    XArchIntrinsicConstants_Avx10v1 = 0x40000,
+    XArchIntrinsicConstants_Avx10v1_V512 = 0x80000,
+    XArchIntrinsicConstants_Evex = 0x100000,
```

Comment:
Do we really need the Evex bit here? It is implied by other bits.

Reply:
The `Evex` bit is intentional and a bit special (will work on getting an explanation of it added to the docs).

The consideration is that with `Avx10v1` we have two different ways to say that certain 128-bit or 256-bit EVEX encoded instructions are supported. These instructions having been made available originally as part of `AVX512VL` (which requires 512-bit support) but now also being available in `AVX10v1` (which does not require 512-bit support).

So, if code is using say `Avx512F.VL.*` (and never uses say `Avx512F.*`), then what we want is to actually track this as requiring the EVEX bit and not as requiring `Avx10v1` or `Avx512` specifically, since it will work provided that either of these two ISAs are supported. This allows the JIT to track and report the actual requirements for certain instructions more cleanly (we only need to check the `EVEX` bit for such instructions) and allows NAOT images to be generated that don't fail if you're code was written using one but was generated using another.

If we had known about `Avx10v1` prior to shipping `Avx512` support, I imagine we would've exposed the managed API surface differently. We probably would've intentionally designed the surface area a bit differently (even with it not matching hardware). In particular, the actual setup we have is basically the following, where `AvxF+BW+CD+DQ` and below that have a nested `V512` type representing the 512-bit extensions and where we would've come up with an ISA name to represent `AvxF+BW+CD+DQ` (possibly just `Evex`, maybe something different):
```mermaid
graph TD;
    Sse2-->Aes;
    Vex-->Avx;
    Avx-->Avx2;
    Evex-->AvxF+BW+CD+DQ;
    AvxF+BW+CD+DQ-->AvxVbmi;
    AvxF+BW+CD+DQ-->AvxIfma;
    AvxF+BW+CD+DQ-->AvxBf16;
    AvxF+BW+CD+DQ-->AvxBitalg;
    AvxF+BW+CD+DQ-->AvxFp16;
    AvxF+BW+CD+DQ-->AvxVbmi2;
    AvxF+BW+CD+DQ-->AvxVpopcntdq;
    AvxVbmi-->Avx10v1;
    AvxIfma-->Avx10v1;
    AvxBf16-->Avx10v1;
    AvxBitalg-->Avx10v1;
    AvxFp16-->Avx10v1;
    AvxVbmi2-->Avx10v1;
    AvxVpopcntdq-->Avx10v1;
    Avx2-->AvxVnni;
    Vex-->Bmi1;
    Vex-->Bmi2;
    Avx2-->Evex;
    Fma-->Evex;
    Avx-->Fma;
    X86Base-->Lzcnt
    Sse2-->Pclmulqdq
    Sse42-->Popcnt
    X86Base-->Sse;
    Sse-->Sse2;
    Sse2-->Sse3;
    Sse3-->Ssse3;
    Ssse3-->Sse41;
    Sse41-->Sse42;
    Sse42-->Vex;
```

Reply:
I think it is fine for the JIT to have internal methods that return true when it is fine to use Vex or Evex instruction encodings. I do not see why the Evex bit has to leak outside the JIT all the way to the minipal. The Vex bit does not leak all the way to the minipal either.

Reply:
The consideration is that for `EVEX` it's not internal to the JIT.

Anytime the JIT does a `compExactlyDependsOn(InstructionSet_*)` or `compOpportunisticallyDependsOn(InstructionSet_*)` query, it gets reported back to the VM. The different VMs handle it as follows:
* For JIT this reporting is a `no-op`
* For R2R this reporting exactly tracks the `InstructionSet_*` queried and whether it was "exactly" or "opportunistic"
  * This is where the cleanup of `InstructionSet_*` would be beneficial, because we have lots more flags than actually needed
* For AOT this reporting gets compressed back to `XArchIntrinsicConstants` which gets queried for an exact match on AOT startup

So, we need the JIT query for `EVEX` usage to map back to `InstructionSet_EVEX` for R2R and `XArchIntrinsicConstants_EVEX` for AOT that way if the code only uses `Avx512F.VL.Abs`, it can work when `Avx10v1` without V512 support exists. Otherwise we'd need it to report as `Avx10v1` or `Avx512`, both of which cut out hardware that would otherwise work with the binary. We only query the more exact ISA when it isn't part of that synthetic `EVEX baseline` and thus needs to restrict the target.

For `VEX` there isn't any need, because the hierarchy is logical and makes sense. There is no historical split where two separate ISAs give access to the same underlying instructions.

Reply:
Ok, got it. 

EVEX is low-level instruction encoding. I think it would be best to confine it to the JIT. The mapping between EVEX and the ISAs is a bit of mental leap outside the JIT.

Would it be better to have `Avx512_VLOnly` ISA that is the common subset of `Avx512` and `Avx10v1`? I assume that all `Avx512*.VL` ISAs are supported in `Avx10v1`. Is that correct?

> if the code only uses Avx512F.VL.Abs, it can work when Avx10v1 without V512 support exists.

Is the user code going to look like this:
```
if (Avx512F.VL.IsSupported || Avx10v1.IsSupported)
{
    ... code that uses Avx512F.VL.Abs ...
}
```
Or are we going to make `Avx512F.VL.IsSupported` return `true` for `Avx10v1` so that `Avx10v1.IsSupported` part of the condition is not needed?

Reply:
> I assume that all Avx512*.VL ISAs are supported in Avx10v1. Is that correct?

Yes, effectively. `Avx10v1` is the "converged" ISA as the official documentation refers to it and unifies almost all the instruction sets exposed so far.

There is some minor nuance in that some ISAs like `VAES`, `VPCLMULQDQ`, and `GFNI` continue to be fully independent and enumerated by their own unique CPUID. This is primarily because they are more specialized scenarios or related to concepts like cryptography and so aren't guaranteed to always be present in all processors.

> Would it be better to have bits for Avx512_VLOnly virtual ISA that is the common subset of Avx512 and Avx10v1?

I'm not sure that's a more clear name. I'll try to brainstorm on some alternatives that are distinct from `EVEX` and see if the interested parties can come to an agreement on something sensible.

Maybe, just as an idea, we could call it just `Avx10` or `Avx10v0` (representing the baseline common to `Avx512` and `Avx10v1`)? Alternatively we could call it just `AvxF_BW_CD_DQ`?
-- This latter is roughly the split Intel is already using. They have, for example, `Avx512Vnni` and `AvxVnni` for the 128 and 256-bit support that exists independently of 512-bit support.

> Or are we going to make Avx512F.VL.IsSupported return true for Avx10v1 so that Avx10v1.IsSupported part of the condition is not needed?

Due to how we've documented the `IsSupported` flags to work up until this point, a user querying that `Avx512F.VL.IsSupported`  is true can assume that it means `Avx512F.IsSupported` is also true, so we'd need the user to query for `Avx512F.VL` and `Avx10v1`.
-- This is where had we known `Avx10v1` was coming prior to implementing `Avx512`, we'd have exposed it differently to make the exposed surface area more "unified" and cleaner for everyone to interact with.

The common baseline between `Avx10v1` and `Avx512` is `128-bit` and `256-bit` encodings for the `F+BW+CD+DQ` ISAs. Then there is additional ISAs in the form of `VBMI`, `IFMA`, `VNNI`, `BF16`, `VPOPCNTDQ`, `VBMI2`, `BITALG`, and `FP16` which are not part of our `Avx512` baseline but which are part of the `Avx10v1` baseline. Finally there is the `VAES`, `GFNI`, and `VPCLMULDQD` ISAs which aren't part of either baseline

So in an ideal world we'd have modeled this effectively as (pseudo-code and maybe with better names):
```csharp
// represents F+BW+CD+DQ
interface Evex : Avx2, Fma { interface V512 { } }
interface AvxVbmi : Evex { interface V512 : Evex.V512 { } }
interface AvxIfma : Evex { interface V512 : Evex.V512 { } }
interface AvxVnni : Evex { interface V512 : Evex.V512 { } }
interface AvxBf16 : Evex { interface V512 : Evex.V512 { } }
interface AvxVpopcntdq : Evex { interface V512 : Evex.V512 { } }
interface AvxVbmi2 : Evex { interface V512 : Evex.V512 { } }
interface AvxBitalg : Evex { interface V512 : Evex.V512 { } }
interface AvxFp16 : Evex { interface V512 : Evex.V512 { } }
interface Avx10v1 : AvxVbmi, AvxIfma, AvxVnni, AvxBf16, AvxVpopcntdq, AvxVbmi2, AvxBitalg, AvxFp16
{
    interface V512 : AvxVbmi.V512, AvxIfma.V512, AvxVnni.V512, AvxBf16.V512, AvxVpopcntdq.V512, AvxVbmi2.V512, AvxBitalg.V512, AvxFp16.V512 { }
}
```

-- On a slightly related point, I believe it would it be a binary breaking change to change from an `abstract class` to an `interface` (i.e. change `public abstract class Avx` to instead be `public interface Avx`), correct? Doing so would allow us to properly model the multiple inheritance that exists here, which might in turn let us do something a little bit clever here to make this "nicer". In particular, we could define the actual `Avx10v1` hierarchy like this and make the existing `Avx512F.VL` inherit from `AvxF` and `AvxF.V512` inherit from `Avx512F`, allowing for everything to be "properly" modeled with the correct implications. Not quite as cleanly as if we had done it from day 1, but more sensibly than what it right now. But my expectation is it would cause downstream issues/breaks and so isn't an option.

Reply:
> Maybe, just as an idea, we could call it just Avx10 or Avx10v0 (representing the baseline common to Avx512 and Avx10v1)? Alternatively we could call it just AvxF_BW_CD_DQ?

I have looked around how other projects are solving this naming problem. [LLVM](https://github.com/llvm/llvm-project/blob/main/clang/docs/UsersManual.rst#x86) seems to use `EVEX256` name for the `Avx10v1` and `Avx512` intersection. I guess it is how we ended up with `EVEX` here. I like the `256` suffix. It helps with clarifying that the virtual ISA represents the non-512 intersection.

> On a slightly related point, I believe it would it be a binary breaking change to change from an abstract class to an interface (i.e. change public abstract class Avx to instead be public interface Avx), correct? 

It should not be a breaking change for typical uses, but as you have said there are likely cases that it would break.

I am not sure how well the ISAs modeling via inheritance hierarchy is going to work going forward. I would not be surprised if we see more places where new hardware revisions carve out subsets of existing instructions and give it a new ISA name. It just happened with Avx512/Avx10v1, Arm64 streaming mode is heading that way, and RISCV may end up there too given its development process. We may want to think about different approach to accommodate these trends.

Reply:
> I guess it is how we ended up with EVEX here. I like the 256 suffix. It helps with clarifying that the virtual ISA represents the non-512 intersection.

👍 

> I am not sure how well the ISAs modeling via inheritance hierarchy is going to work going forward.

I agree we should think some more, but overall I do believe this is still the correct model. We've seen quite a bit of praise on how we've modeled everything as compared to how C/C++ or Rust do it. The overall separation of `xplat APIs` vs `platform specific APIs` and then the division of the latter into clear classes that are based on the underlying hardware checks has helped make discoverability, approachability, and usability of the feature in .NET one of the best.

> It just happened with Avx512/Avx10v1

This one I think was more x64 having tried something new with `Avx512` and then finding out that it didn't work in practice, so they are now fixing it by reconverging everything with `Avx10v1`. The architectural documents around `Avx10v1` then cover how they plan to handle this moving forward and that it's effectively going back to the model we had seen previously with `Sse` and `Avx` (where you have a baseline encoding and then incremental functionality extended on top of this in a version like manner).

> Arm64 streaming mode is heading that way

This is a case where there is still a clear hierarchy given we have `SveStreaming` and `SveNonStreaming` as the split.

Even if we had shipped `class Sve` prior to streaming support becoming defined, we could have inserted a `class SveStreaming` and had `Sve` derive from it, moving relevant methods down (since that is safe to do for classes).

Reply:
-- What I'll plan on doing given the conversation so far is getting this in as is first as it fixes some bugs without the refactoring being too complex and puts us in a good position for .NET 9.

I'll then follow up with a PR that does a broader cleanup of the data used by R2R and changes `EVEX` to `EVEX256` (which will itself require some refactoring in the JIT).

Reply:
> streaming and non-streaming support do not form a hierarchy

I think this is more just the names aren't the "right" ones.

What I had meant is that you have two sets of instructions. That is, there are SVE instructions that are available both in regular SVE and in SVE streaming mode. Then there are SVE instructions which are not available in streaming mode unless another architectural bit is set.